src/ZkFold/Base/Protocol/Protostar/Fold.hs

```diff
@@ -0,0 +1,179 @@
+
+type FS_CM n c a = FiatShamir a (CommitOpen a c (RecursiveCircuit n a))
+
+transform
+    :: forall n c a
+    .  HomomorphicCommit a [Map Natural a] c
+    => a
+    -> RecursiveCircuit n a
+    -> Vector n a
+    -> FS_CM n c a
+transform ck rc v = FiatShamir (CommitOpen (hcommit ck) rc) v
+
+-- | These instances might seem off, but accumulator scheme requires this exact behaviour
+--
+instance AdditiveSemigroup a => AdditiveSemigroup (Map Natural a) where
+    (+) = M.unionWith (+)
+
+instance AdditiveMonoid a => AdditiveMonoid (Map Natural a) where
+    zero = M.empty
+
+instance (Ring a, P.Eq a) => Acc.LinearCombination (Map Natural a) (Map Natural (PU.Poly a)) where
+    linearCombination mx ma = M.unionWith (+) (PU.monomial 1 <$> mx) (PU.constant <$> ma)
+
+instance (Ring a, P.Eq a, KnownNat n) => Acc.LinearCombination (Vector n a) (Vector n (PU.Poly a)) where
+    linearCombination mx ma = (+) <$> (PU.monomial 1 <$> mx) <*> (PU.constant <$> ma)
+
+instance (Ring a, KnownNat n) => Acc.LinearCombinationWith a (Vector n a) where
+    linearCombinationWith coeff a b = (+) <$> (P.fmap (coeff *) a) <*> b
+
+fold
+    :: forall a n c x
+    .  Arithmetic a
+    => x ~ ArithmeticCircuit a (Vector n)
+    => P.Eq c
+    => Scale a c
+    => Scale a a
+    => AdditiveGroup c
+    => RandomOracle a a
+    => RandomOracle c a
+    => RandomOracle [c] a
+    => HomomorphicCommit a [a] c
+    => HomomorphicCommit a [Map Natural a] c
+    => KnownNat n
+    => (Vector n (FieldElement x) -> Vector n (FieldElement x))  -- ^ An arithmetisable function to be applied recursively
+    -> Natural                             -- ^ The number of iterations to perform
+    -> SPS.Input a (RecursiveCircuit n a)  -- ^ Input for the first iteration
+    -> FoldResult n c a
+fold f iter i = foldN iter ck rc i [] initialAccumulator (Acc.KeyScale one one)
+    where
+        rc :: RecursiveCircuit n a
+        rc = RecursiveCircuit iter (compile @a f)
+
+        initE = hcommit ck $ replicate (SPS.outputLength @a rc) (zero :: a)
+
+        ck = oracle i
+
+        initialAccumulator :: Accumulator (Vector n a) a c (Map Natural a)
+        initialAccumulator = Accumulator (AccumulatorInstance (P.pure zero) [hcommit ck [zero :: Map Natural a]] [] initE zero) [zero]
+
+
+instanceProof
+    :: forall n c a
+    .  Arithmetic a
+    => KnownNat n
+    => Scale a a
+    => HomomorphicCommit a [Map Natural a] c
+    => a
+    -> RecursiveCircuit n a
+    -> SPS.Input a (RecursiveCircuit n a)
+    -> InstanceProofPair (Vector n a) c (Map Natural a)
+instanceProof ck rc i = InstanceProofPair i (NARKProof [hcommit ck [m]] [m])
+    where
+        m = SPS.prover @a rc M.empty i []
+
+foldN
+    :: forall n c a
+    .  Arithmetic a
+    => P.Eq c
+    => AdditiveGroup c
+    => Scale a c
+    => Scale a a
+    => KnownNat n
+    => RandomOracle a a
+    => RandomOracle c a
+    => RandomOracle [c] a
+    => HomomorphicCommit a [a] c
+    => HomomorphicCommit a [Map Natural a] c
```

Comment:
These look like references to concrete instances, maybe it is better to inline them then?